share/doc/wake/basic_wake_tutorial.md

```diff
@@ -0,0 +1,809 @@
+```
+
+### Wake Built-In Functions
+Few of the commonly used built-in functions in this tutorial is given below
+#### Operation on String
+* `explode` breaks a string into a list of string character.
+* `strlen` returns the length of the given string.
+* `cat` concatenates the elements in list of string to a string. 
+#### Operation on List
+* `take` takes first `n` elements in the list.
+* `drop` drops first `n` elements in the list.
+* `reverse` reverses the elements in the list.
+* `head` returns the first element in the list.
+* `tail` returns the list by removing the `head`.
+* `prepend` places the given data at the begining(head) of the given list.
+* `append` appends the given data at the end of the given list.
+* `++` concatenates any number of list into a single list.
+* `len` returns the length of the given list.
+
+* Function duplicate concantenates or copies a given string twice
+```
+def duplicate s  ="{s}{s}"
+```
+
+* integerToUnicode is a `wake built-in function` which takes the ASCII input and provides its equivalent character
+```
+wake -x 'integerToUnicode 99'
+```
+
+* unicodeToInteger is a `wake built-in function` which gives the ASCII value of the given character.
+```
+wake -x 'unicodeToInteger "a"'
+```
+
+* strlen is a `wake built-in function` which returns the length of the given string.
+```
+wake -x 'strlen "abcde"'
+```
+
+### List
+A List is a sequence of items of same type. Empty list returns `Nil`. `head` returns the first element in the list. `tail` returns the rest of the list without the `head`.
+
+* Example creates a list of given elements
+```
+def listcreate xx = xx,xx,xx,Nil
+```
+ 
+* Example shows how to prepend the list of integers with the given value
+```
+def d = 1,2,3,Nil
+def listprepend data1 = prepend data1 d
+```
+
+* concantenates 2 lists using ++ operator
+```
+wake -x '(1,2,3,Nil) ++ (4,5,6,Nil)'
+```
+
+### Type Inference
+* Here the paramater c has to be string since function strlen as been applied to it, and b has to be integer as it is being added to the output of `strlen c` and a is list of integers which prepends the output of `(b + strlen c)`.
+```
+def madeup a b c =  (b + strlen c), a 
+```
+
+#### Parenthesis vs Pipe
+* `|`is the pipeFn which takes the argument from the left hand side of the `|` and provides it to the right hand side function. This makes the function to be more readable. `(a(b(c(d))))` is represented as `d | c | b | a`. In both the cases d is evaluated first then followed by c , b and a.
+```
+def reversestring s = cat (reverse (explode (s)))
+```
+
+The same can be expressed with the pipe function as below
+```
+def reversestring s = s | explode | reverse | cat
+```
+
+* Below function takes a string and returns the middle character of the string.
+```
+def middle fullstring =
+  def slength = if (strlen fullstring % 2 == 0) then (strlen fullstring / 2) else  ((strlen fullstring / 2) + 1)
+  def outputstr = explode fullstring | take slength | reverse | head 
```

Comment:
it seems like back in the `def` functions section we should explain that you can arbitrarily deeply put `def` inside of other `def`. And demonstrate that you can't "see" the `def` outside the appropriate name space.

It might be nice to show the kinds of error messages people will encounter if they do such a thing

Reply:
Added the explnation